File: MATLAB Codes/WP_Energy_Ratio.m
function fig = WP_Energy_Ratio(L, stickout_length,list_name,ts_no,WF,N,row_no,column_no,Fs) 
%% Wavelet Packet Decomposition and Energy Ratio Computation
% Description: This code takes time series of cutting tests and computes 
%              their Wavelet Packet Decomposition. Energy ratios of each 
%              wavelet packet, reconstructed wavelet packets in time domain
%              for each transform and FFT of these packets are plotted as outputs.
%              These plots help user to indentify the informative wavelet
%              packet for given time series. 
% 
% INPUT:
%       L              : Wavelet Packet Decomposition Level
%       stickout_length: Overhang length of the cutting tool 
%       list_name      : .txt file that includes the names of the time 
%                        series for each of the stickout size (overhang length)
%       ts_no          : 1 x n array that includes the number of the time series 
%                      : which user wants to investigate their WPT inside of the list 
%       WF             : Wavelet Function
%       N              : number of time series whose WPT will be computed
%       row_no         : row number for the plots that include reconstructed
%                        wavelet packets in time domanin and frequency domain
%       column_no      : column number for the plots that include reconstructed
%                        wavelet packets in time domanin and frequency domain
%       Fs             : sampling frequency of the time series
%
% OUTPUT: - 1st Plot: It shows energy ratio of each packet of the decomposition 
%         for selected time series inside of the list
%         - 2nd to (N+1)th Plot: Plots of reconstructed packets of WPT for each
%         investigated time series in time domain
%         - (N+2)th to (2N+1)th Plot: FFT of reconstructed packets of WPT for each
%         investigated time series 
%
% Note:   Change the file paths inside of the function to be able to access
%         to cutting test data
%
% References
% 1) M. C. Yesilli, F. A. Khasawneh, and A. Otto, “On transfer learning for chatter detection in turning
% using wavelet packet transform and empirical mode decomposition,” CIRP J. Manuf. Sci. Technol. 2019, in press.
% 2)F. Khasawneh, A. Otto, and M.C. Yesilli, “Turning dataset for chatter diagnosis using machine learning.”
% Mendeley Data, v1. http://dx.doi.org/10.17632/hvm4wh3jzx.1, 2019.


%add paths for the current data folder for cutting tests
if stickout_length==2
    addpath('D:\Data Archive\Cutting_Test_Data_Documented\cutting_tests_processed\2inch_stickout');
elseif stickout_length==2.5
    addpath('D:\Data Archive\Cutting_Test_Data_Documented\cutting_tests_processed\2p5inch_stickout');
elseif stickout_length==3.5
    addpath('D:\Data Archive\Cutting_Test_Data_Documented\cutting_tests_processed\3p5inch_stickout');
elseif stickout_length==4.5
    addpath('D:\Data Archive\Cutting_Test_Data_Documented\cutting_tests_processed\4p5inch_stickout');
end


filename = regexp(fileread(list_name), '\r?\n', 'split');

for i=1:N
    %load mat files
    dataname = sprintf('%s',filename{ts_no(i)});

    % extract the time series information from its name
    rpm(i) = str2num(dataname(3:5));
    doc(i) = str2num(sprintf('0.%s',dataname(7:9)));
    label(i) = string(dataname(1));
    time_series{i}=load(dataname);

    eval(sprintf('ts%d=time_series{1,%d}.tsDS(:,2);',i,i));
    eval(sprintf('time_%d=time_series{1,%d}.tsDS(:,1);',i,i));
    
    % Wavelet Packet Decomposition
    eval(sprintf('WPT%d=wpdec(ts%d,%d,"%s");',i,i,L,WF));
    
    %Energy computation for each packet
    eval(sprintf('E%d=wenergy(WPT%d);',i,i));

    %Energy ratios 
    eval(sprintf('energyratio%d=(E%d)/100;',i,i));
end

%Total Wavelet packet number
WP_No=2^L;

%% Plotting

for i=1:N
    eval(sprintf('bpcombined(:,%d) = transpose(energyratio%d);',i,i));
    dataname = sprintf('%s',filename{ts_no(i)});
    if label(i) == 'c'
        legend_names(i)= string(sprintf('Chatter,Stickout=%d inch, RPM=%d, DOC=%.4f',stickout_length,rpm(i),doc(i)));
    elseif label(i) == 's'
        legend_names(i)= string(sprintf('Stable,Stickout=%d inch, RPM=%d, DOC=%.4f',stickout_length,rpm(i),doc(i)));
    elseif label(i) == 'i'
        legend_names(i)= string(sprintf('Mild Chatter,Stickout=%d inch, RPM=%d, DOC=%.4f',stickout_length,rpm(i),doc(i)));
    end
end

%bar plot of energy ratios
binedges=(1:WP_No);
figure(1);
h = bar(binedges,bpcombined,'grouped');

label_fontsize = 18;
ticks_fontsize = 18;

% add a legend and format it
legend(legend_names);
set(legend, 'interpreter', 'latex', 'FontSize', 15);

% format the ticks
ax=gca();
set(ax, 'TickLabelInterpreter', 'Latex', ...
    'FontSize', ticks_fontsize);

% set x and y labels and formatting
ax.XLabel.String = 'Wavelet Packets';
ax.XLabel.FontSize = label_fontsize;
ax.XLabel.Interpreter = 'Latex';

ax.YLabel.String = 'Energy Ratios';
ax.YLabel.Interpreter = 'Latex';
ax.YLabel.FontSize = label_fontsize;
fig = gcf();
%% Reconstruction of each wavelet packet and their plot

%wavelet packet nodes to be used in the 
terminal_nodes_start = 2^(L)-1;
terminal_nodes_end = 2^(L+1)-2;


for i=1:N
  % array name which will include the reconstructed packets inside for each
  % time series
  a_name = sprintf('recon_%d',i);
  eval(sprintf(' time_vec =time_%d;',i))
  %wavelet packet number defined by matlab wavelet tree
  terminal_node = terminal_nodes_start;
  
  figure(i+1)
  
  for j=1:WP_No
    eval(sprintf('%s{%d} = wprcoef(WPT%d,%d);',a_name,j,i,terminal_node));  
    
    % plotting of each reconstructed packets
    h=subplot(row_no,column_no,j);
    pos1(j,1:4)=get(h,'position');
    eval(sprintf('plot(time_%d,%s{%d})',i,a_name,j))
    xmax=max(time_vec);
    xmin=min(time_vec);
    xlim([xmin,xmax]);
    k=sprintf('Packet %d',j);
    title(k,'interpreter', 'latex', 'FontSize', 12);
    ax=gca();
    set(ax, 'TickLabelInterpreter', 'Latex', ...
        'FontSize', ticks_fontsize);
    eval(sprintf('fig_%d=gcf;',i+1)); 
    terminal_node = terminal_node+1;
  end 
  height = (pos1(1,2)+pos1(1,4)-pos1(end,2));
  h3=axes('position',[pos1(1,1)-0.03 pos1(end,2) pos1(end,3) height],'visible','off');
  y_label=ylabel('Amplitude $(\frac{m}{s^2})$','visible','on','interpreter', 'latex', 'FontSize', 20);
  h4=axes('position',[pos1(1,1)+0.31 pos1(end,2)-0.01 pos1(end,3) height],'visible','off');
  x_label=xlabel('Time(s)','visible','on','interpreter', 'latex', 'FontSize', 20);
end
%% FFT for Reconstructed Signals
for i=1:N
    eval(sprintf(' time_vec =time_%d;',i))
    figure(i+N+1)
    for j=1:WP_No
        % define the signal parameters
        T = 1/Fs;             % Sampling period 

        eval(sprintf('time_series = recon_%d{%d};',i,j))
        l = length(time_series); % number of points in the signal

        X=time_series;

        % take the fft of the signal
        NFFT = 2^nextpow2(l);  % Next power of 2 from length of X
        Y = fft(X, NFFT);
        % number of unique frequency points
        NumUniquePts = ceil((NFFT+1)/2); 
        % grab only the unique components of Y
        Y = Y(1:NumUniquePts);
        MX=abs(Y)/l;  % take magnitude of Y and divide by L to get the linear spectral density

        % multiply by 2 to account for throwing out the second half of the spectrum
        MX = MX * 2;
        % However, the DC component of the signal is unique and should not be 
        %multiplied by 2:
        % Y(1) is the DC component of x for both ODD and EVEN NFFT.
        % Also, for odd NFFT, the Nyquist frequency component is not evaluated, 
        % and the number of unique points is (NFFT+1)/2.
        MX(1) = MX(1)/2; 
        % If NFFT is even, then the FFT will be symmetric such that 
        % the first (1+NFFT/2) points are unique, and FFTX(1+NFFT/2) is 
        % the Nyquist frequency component of x, which is unique. The rest
        % are symmetrically redundant.
        % when NFFT is even, FFTX(1+NFFT/2) is the Nyquist 
        %frequency component of x, and we need to make sure it is unique.
        if ~rem(NFFT, 2)  
             MX(length(MX))=MX(length(MX))/2;
        end 
        f = Fs/2*linspace(0, 1,  NumUniquePts);
        %plotting
        h2=subplot(row_no,column_no,j);
        pos2(j,1:4)=get(h,'position');
        plot(f, MX)
        %formatting the figure
        xmax=max(f);
        xmin=min(f);
        xlim([xmin,xmax])
        k=sprintf('Packet %d',j);
        title(k,'interpreter', 'latex', 'FontSize', 12)
        ax=gca();
        set(ax, 'TickLabelInterpreter', 'Latex', ...
            'FontSize', ticks_fontsize);
        eval(sprintf('fig_%d=gcf;',i+1+N));
    end
    height = (pos1(1,2)+pos1(1,4)-pos1(end,2));
    h5=axes('position',[pos2(1,1)-0.65 pos2(end,2) pos2(end,3) height],'visible','off');
    y_label=ylabel('$|X(f)|$','visible','on','interpreter', 'latex', 'FontSize', 20);
    h6=axes('position',[pos2(1,1)-0.32 pos2(end,2)-0.01 pos2(end,3) height],'visible','off');
    x_label=xlabel('Frequency(Hz)','visible','on','interpreter', 'latex', 'FontSize', 20);    
end
end

File: WPT_EEMD_ML/WP_Energy_Ratio.py

import numpy as np
import pywt
import scipy.io as sio
import matplotlib.pyplot as plt
from scipy.fftpack import fft
from matplotlib import rc
rc('font',**{'family':'sans-serif','sans-serif':['Helvetica']})
rc('text', usetex=True)
plt.rc('text', usetex=True)
plt.rc('font', family='serif')


def wEnergy(coeffs):
    
    '''
    This function takes wavelet coefficients and return the energy ratio of wavelet packets.
    
    :param list (coeffs): The list of wavelet coefficients obtained from Wavelet Packet Transform
    
    :Returns:
        (np.array([])) The energry ratios of each wavelet packets 
    
    
    '''
  
    Et = 0
    energy = np.zeros(len(coeffs))
    for k in range(len(coeffs)):
        energy[k] = sum(coeffs[k]**2)
        Et = Et + sum(coeffs[k]**2)
        
    en_ratio = energy/Et
    
    return en_ratio

def AbsFFT(ts, fs):

    '''
    This function computes the normaliez spectrum of the given 1D signal.
    
    :param array (ts): time series
    
    :param int (fs): sampling frequency
    
    :Returns:
        :xf:
            (np.array([])) frequency values
        
        :yf:
            (np.array([])) amplitudes
          
    '''
    fs = fs/2
    ts = ts.reshape(len(ts,))
    t = 1/fs 
    N = len(ts)
    
    #creates array for time data points to associate to time series
    xf = np.split(np.linspace(0.0, 1.0/t, N//2),[1])[1] 
    #converts array of time data points to appropriate frequency range and removes first term
    
    yf = fft(ts) #computes fast fourier transform on time series
    yf = (np.abs(yf[0:N//2])) #converts fourier transform to spectral density scale
    yf = np.split(yf,[1])[1] #removes first term (tends to be infinity or overly large)
    yf = yf/max(yf) #Normalizes fourier transform based on maximum density
    return(xf, yf)

def WP_Energy_Ratio(data_path, list_name, ts_no, WF, L, case_no, fs, plot_recon, layout=None ):
    
    '''
    This function returns plots of energy ratios of wavelet packets, reconstructed signals of each wavelet packet and their spectrum.
    Reconstructed signals is plotted when plot_recon is set to 'True'.
    
    :param str (data_path): Path to folder where list of time series names and time series data is kept
    
    :param str (list_name): Name of the .txt file which includes names of time series data
    
    :param list (ts_no): Index of time series which user wants to investigate
    
    :param str (WF): Wavelet function (see types_ of wavelet functions available in PyWavelet package)

    :param int (L): Transformation level 
    
    :param int (case_no): Overhang distance 
    
    :param int (fs): Sampling frequency
    
    :param bool (plot_recon): It is set to 'True' to obtain the reconstructed signals plots 
    
    :param list (layout): Row and column numbers for the plot of reconstructed signals 
    
    .. _types: https://pywavelets.readthedocs.io/en/latest/regression/wavelet.html?highlight=wavelet%20list#wavelet-families-and-builtin-wavelets-names
        
    :Returns:
        :figures:
            First figure is the energy ratio plot. If user set to plot_recon to 'True', then algorithm returns reconstruted signals figures as well.
            For instance, if the number of time series that user wants to investigate is 3, and algorithm returns two figures for each time series.
            Totally, it returns 7 figures. 
            Plots of reconstructed signals and their spectrums are given in the order user specified as input with ts_no.
        
          
    '''
    
    # font size for the labels in plots
    f_s = 15
    
    # file path to folder where the data is kept
    file_path = data_path+'\\'+list_name
    
    # number of time series which will be investigated 
    N = len(ts_no)
    
    # read the file that includes the name of datafiles
    with open(file_path) as f:
        data_names = f.read().splitlines() 
      
    rpm = []
    doc = []
    label = []
    legend_name=[]
    ts = np.zeros((N),dtype=object)
    ratio = np.zeros((N,2**L))
    recon = np.zeros((N,2**L),dtype=object)
    
    # compute wavelet coefficients and energy ratios
    for i in range(N):
        # load mat files
        name = data_names[ts_no[i]-1]
        
        # time series information
        rpm.append(str(name[2:5]))
        doc.append(str('0.'+name[6:9]))
        
        if name[0]=='c':
            label.append('Chatter')
        elif name[0]=='s':
            label.append('Stable')
        elif name[0]=='i':
            label.append('Mild Chatter')
            
        legend_name.append(label[i]+', Stickout='+str(case_no)+' inch, RPM='+str(name[2:5])+', DOC='+str('0.'+name[6:9]))
        
        #load the time series
        ts[i] = sio.loadmat(data_path+'\\'+name)['tsDS']
        
        # apply wavelet packet decomposition
        wp = pywt.WaveletPacket(data=ts[i][:,1], wavelet=WF,maxlevel=L,mode='symmetric')
        packet_order = [node.path for node in wp.get_level(L, 'freq')]
        coeffs = [wp[path].data for path in packet_order]
        
        #reconstruct the wavelet packets
        for j in range(2**L):
            # create a empty wavelet packet object
            new_wp = pywt.WaveletPacket(data=None, wavelet=WF, mode='symmetric')
            new_wp[packet_order[j]] = wp[packet_order[j]]
            recon[i,j] = new_wp.reconstruct(update=False)
            # find the diff between reconstructed length and time series length
            len_dif = len(ts[i])-len(recon[i,j])
            recon[i,j] = recon[i,j][:len_dif]
            
        #energy ratio
        ratio[i,:] = wEnergy(coeffs)
        
        
    # plot energy ratios   
    barWidth = 0.25
    fig = plt.subplots(figsize =(12, 5)) 
       
    for i in range(N):
        # Set position of bar on X axis
        if i == 0:
            br = np.arange(2**L)+1
        else:
            br = [x + barWidth for x in br]
            
        plt.bar(br, ratio[i,:], width = barWidth, label = legend_name[i])
    
    plt.xlabel('Wavelet Packets' , fontsize = f_s)
    plt.ylabel('Energy Ratios', fontsize = f_s)    
    plt.legend()  
    plt.subplots_adjust(hspace=0.5,wspace=0.5)
    ax=plt.gca()
    ax.tick_params(axis='both', labelsize=f_s-1)
    
    # plot reconstructed signals and their spectrum
    if plot_recon:
        for i in range(N):
            fig, axes = plt.subplots(nrows=layout[0], ncols=layout[1],figsize=(12,9)) 
            fig_FFT, axes_FFT = plt.subplots(nrows=layout[0], ncols=layout[1],figsize=(12,9))
             
            axes = axes.flatten()
            axes_FFT = axes_FFT.flatten()
    
            
            for j in range(2**L):   
                ax = axes[j]
                ax.plot(ts[i][:,0],recon[i,j])
                ax.set_xlabel('time (seconds)' , fontsize = f_s)
                ax.set_ylabel('Amplitude', fontsize = f_s)  
                title = 'Packet'+str(j+1)
                ax.set_title(title)
                ax.tick_params(axis='both', labelsize=f_s-1)
                
                plt.subplots_adjust(hspace=1 ,wspace=1)
                
                ax = axes_FFT[j]
                xf, yf = AbsFFT(recon[i,j], fs)
                ax.plot(xf,yf)
                ax.set_xlabel('frequency (Hz)' , fontsize = f_s)
                ax.set_ylabel('$\mid X(f) \mid $', fontsize = f_s)  
                title = 'Packet'+str(j+1)
                ax.set_title(title, fontsize = f_s)
                ax.tick_params(axis='both', labelsize=f_s-1)
                
                plt.subplots_adjust(hspace=0.8,wspace=0.5)
                